service/matching/backlog_manager.go

```diff
@@ -298,3 +299,35 @@ func (c *backlogManagerImpl) newIOContext() (context.Context, context.CancelFunc
 func (c *backlogManagerImpl) queueKey() *PhysicalTaskQueueKey {
 	return c.pqMgr.QueueKey()
 }
+
+// countTasksFromTaskQueue counts the total number of tasks present in a task queue; returns 0 for
+// cassandra databases since the in-memory counter is a good estimate for the number of tasks present while it returns
+// the actual number of tasks in the task queue for sql databases. This is required since SQL databases do not persist
+// taskQueueInfo when tasks are created.
+func (c *backlogManagerImpl) countTasksFromTaskQueue(ctx context.Context) (int64, error) {
+	tasksPresent, err := c.db.store.CountTasksExact(ctx, &persistence.CountTasksExactRequest{
+		NamespaceID: c.db.queue.NamespaceId().String(),
+		TaskQueue:   c.db.queue.PersistenceName(),
+		TaskType:    c.db.queue.TaskType(),
+	})
+	if _, ok := err.(*serviceerror.Unimplemented); ok {
+		tasksPresent = 0
+	} else if err != nil {
+		return 0, err
+	}
+	return int64(tasksPresent), nil
+}
+
+func (c *backlogManagerImpl) getApproximateBacklogCount(ctx context.Context) (int64, error) {
+	exactTasks, err := c.countTasksFromTaskQueue(ctx)
+	if err != nil {
+		return 0, err
+	}
+
+	if exactTasks > 0 {
```

Comment:
what if exactTask is 0 but it's 0 for real? Probably you want to merge these two functions and rely on Unimplemented error instead of 0 value.